src/graph.rs

```diff
@@ -0,0 +1,343 @@
+    }
+}
+
+/// A node in a graph
+pub struct GraphNode {
+    outgoing_edges: SmallVec<[(GraphNodeID, Edge); 8]>,
+    /// The set of attributes associated with this graph node
+    pub attributes: Attributes,
+}
+
+impl GraphNode {
+    fn new() -> GraphNode {
+        GraphNode {
+            outgoing_edges: SmallVec::new(),
+            attributes: Attributes::new(),
+        }
+    }
+
+    /// Adds an edge to this node.  There can be at most one edge connecting any two graph nodes;
+    /// the result indicates whether the edge is new (`Ok`) or already existed (`Err`).  In either
+    /// case, you also get a mutable reference to the [`Edge`][] instance for the edge.
+    pub fn add_edge(&mut self, sink: GraphNodeRef) -> Result<&mut Edge, &mut Edge> {
+        let sink = sink.0;
+        match self
+            .outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+        {
+            Ok(index) => Err(&mut self.outgoing_edges[index].1),
+            Err(index) => {
+                self.outgoing_edges.insert(index, (sink, Edge::new()));
+                Ok(&mut self.outgoing_edges[index].1)
+            }
+        }
+    }
+
+    /// Returns a reference to an outgoing edge from this node, if it exists.
+    pub fn get_edge(&self, sink: GraphNodeRef) -> Option<&Edge> {
+        let sink = sink.0;
+        self.outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+            .ok()
+            .map(|index| &self.outgoing_edges[index].1)
+    }
+
+    /// Returns a mutable reference to an outgoing edge from this node, if it exists.
+    pub fn get_edge_mut(&mut self, sink: GraphNodeRef) -> Option<&mut Edge> {
+        let sink = sink.0;
+        self.outgoing_edges
+            .binary_search_by_key(&sink, |(sink, _)| *sink)
+            .ok()
+            .map(move |index| &mut self.outgoing_edges[index].1)
+    }
+}
+
+/// An edge between two nodes in a graph
+pub struct Edge {
+    /// The set of attributes associated with this edge
+    pub attributes: Attributes,
+}
+
+impl Edge {
+    fn new() -> Edge {
+        Edge {
+            attributes: Attributes::new(),
+        }
+    }
+}
+
+/// A set of attributes associated with a graph node or edge
+pub struct Attributes {
+    values: SmallVec<[(Identifier, Value); 8]>,
+}
+
+impl Attributes {
+    /// Creates a new, empty set of attributes.
+    pub fn new() -> Attributes {
+        Attributes {
+            values: SmallVec::new(),
+        }
+    }
+
+    /// Adds an attribute to this attribute set.  If there was already an attribute with the same
+    /// name, replaces its value and returns `Err`.
+    pub fn add<V: Into<Value>>(&mut self, name: Identifier, value: V) -> Result<(), ()> {
+        match self.values.binary_search_by_key(&name, |(name, _)| *name) {
+            Ok(_) => Err(()),
+            Err(index) => {
+                self.values.insert(index, (name, value.into()));
+                Ok(())
+            }
+        }
```

Comment:
1. The docs appear to be incorrect; this seems to return `Err` without replacing the value.
2. As above with the graph itself, this seems hard to use correctly. It doesn’t look like we’d have a well-defined union on values in general, but maybe better semantics in this case would involve replacing the value (as the docs suggest) or unioning the values when well-defined (e.g. `Set`/`Set` values could be unioned)?
3. Again as above, can we factor the binary search so we don’t have to be vigilant about updating every other position as the code changes?
4. I believe `Result<(), ()>` is isomorphic to a boolean; if we want to continue reporting collisions, could we return the result as a boolean instead?

Reply:
>     1. The docs appear to be incorrect; this seems to return `Err` without replacing the value.

Good catch!  I'll fix that.



>     4\. I believe `Result<(), ()>` is isomorphic to a boolean; if we want to continue reporting collisions, could we return the result as a boolean instead?

I went with `Result<(), ()>` because then the caller will have some useful methods on the result that will make handling the two cases more ergonomic.  In particular `Result` works with the `?` operator, but a `bool` doesn't.